### ISSUE TITLE
Twitch live videos not working correctly

### ISSUE THREAD

**Issue by SnakeSosa** [igrigorik/videospeed#584]
Only on live video and not vods the video speed is not able to be adjusted. When pressing the increase speed key you can see it quickly switching to the higher speed but after a few milliseconds it reverts back to default speed(same with ads on a livestream) this wasn't always the case and i think has started in last week or two.

Attached gif is showing the speed identifier where im holding down the increase button and it should be going up by .25

![ezgif-3-ca0a00527d3a](https://user-images.githubusercontent.com/46269039/73888368-6dd82500-4865-11ea-8eab-60dd57681537.gif)

**Comment by ChadBailey** [comment]
I must be missing something here, how could one watch a live broadcast at faster than 1x speed? This feels like an intentional design choice targeted at preventing running out of buffer.

**Comment by jedawson** [comment]
i too would like the ability to see into the future. /sarcasm

**Comment by SnakeSosa** [comment]
some times i need to build or run down the buffer depending on my internet as it can be bad but also i've found on twitch that the playback will jump to 1.03 if theres some buffering and because none of the speed increase/decreases aren't working the revert speed function wont work either and i have to reload the entire page

**Comment by jedawson** [comment]
how much time are you leaving in the buffer? i feel like this issue won't really make sense to fix since it is working on the videos on demand on twitch. can't you just wait for them to post it and then watch it later?

**Comment by SnakeSosa** [comment]
sometimes i wont to run down the buffer to like 5 secs or less to reduce latency, cause at times i want to watch live to be able to interact, i know it wouldn't be a major problem to fix but better to note it and be aware of it

**Comment by jedawson** [comment]
well if you have waited a long time for the buffer to increase, you won't be able to interact.

**Comment by ChadBailey** [comment]
> some times i need to build or run down

By virtue of me misunderstanding what you meant by this (my mind thought, "oh, he needs to pause the video to go downstairs") I think I may understand a use-case for adjusting speed on live video feeds. If you have to pause a live stream, there is really no feasible way to catch back up without adjusting speed as you will either have to skip what you missed or remain that far behind the rest of the broadcast.

This may be fixable with my [experimental feature](https://github.com/igrigorik/videospeed/pull/547) if you would like to give it a try.

**Comment by SnakeSosa** [comment]
> > some times i need to build or run down
> 
> By virtue of me misunderstanding what you meant by this (my mind thought, "oh, he needs to pause the video to go downstairs") I think I may understand a use-case for adjusting speed on live video feeds. If you have to pause a live stream, there is really no feasible way to catch back up without adjusting speed as you will either have to skip what you missed or remain that far behind the rest of the broadcast.
> 
> This may be fixable with my [experimental feature](https://github.com/igrigorik/videospeed/pull/547) if you would like to give it a try.

Yep thats one reason i would use it for but twitch can be annoying that if you go too far behind it will just skip forward in the buffer and be back near the actual live-stream.

Just ran it added twitch.tv to the list and yep it works it allows me again to adjust the speed of live-streams, i then retested with the latest version again just to make sure and the problem is still there so your bugfix definitely solves this problem thank you very much!

**Comment by ChadBailey** [comment]
Good to know, seems like this may warrant inclusion in the main branch (as you can see in the discussion, we have been going back and forth on whether to implement/how). Since twitch is probably one of the most watched video platforms this may impact how we do that.

Thanks for testing and confirming!

**Comment by SnakeSosa** [comment]
Yeah was reading into your discussion and I think it will need to become necessary now for me while I know ads pay the bills but some ads are just annoying or can be completely out of touch with issues and people. With the implementation twitch had done it meant the adds couldn't be sped up and I could see others possibly using this in the future to stop ad skippers.

Glad I was able to help

**Comment by ChadBailey** [comment]
Something I want to make clear is I do not support using this extension as a means to skip ads. If you're watching content at 2x and you go through an ad at 2x speed, then I think that's fine. On the other hand, to just use this in place of an ad blocker I think is a very bad idea. There's a couple of reasons I believe this.

First, the act of speeding up video playback is a valid and reasonable thing to do - but skipping advertisements that you're intended to watch as compensation for the service provided is not the intention of VSC. If you read the readme, that's not listed anywhere as a benefit.

Second, I do not want the inevitable outcome of this. That inevitable outcome is websites trying to find countermeasures for altering playback rates. If you want to skip ads, please find another way to do that as I don't want to be on an Internet where I am unable to watch at the speeds I find comfortable.

Unfortunately, this comment very much makes me wary of implementing the fix. I think we may need to open a separate discussion about how we can make sure owners of websites understand this isn't the intention, and what we can do to help cooperate with websites wishing to prevent this kind of abuse.

Please understand if we allow the main use case of this tool to be skipping ads we invite very strong countermeasures like browsers removing the ability to adjust playback rate altogether. I know it may sound extreme now, but this comes ahead of a far-future discussion of this extension going extinct and someone commenting "I never thought it would come to this". Well, now you're thinking about it so you won't be able to make such a claim.

**Comment by SnakeSosa** [comment]
Oh I certainly don't advocated the skipping of ads whole-heartily as it is needed to provide the content we want otherwise we'd have nothing to watch. I guess continued discussion on this may end up leading to people wanting to use it for skipping ads as the lazy option. So either release it silently or don't at all. 

I was thinking about how some creators I know sometimes support ad skipping methods as long as another revenue creating method is used but this being an extension used everywhere and not unique to a creator I don't see as a possibility as many would just take the lazy option and as you said could end up with the removal as you said.

**Comment by ChadBailey** [comment]
Sorry of my words came across as confrontational. I suppose I just wanted to make my stance on that very clear since I consider it pivotal to the long term success of this extension.

**Comment by SnakeSosa** [comment]
No not all, I completely understand and I wouldn't want a useful and helpful extension (that i'm sure a lot of hard work has been put into) being removed.

**Comment by ChadBailey** [comment]
I am working on re-implementing #547 using the new code structure (VSC was changed quite a bit to support videos nested in shadowroots)

In the interim, I recommend you all add twitch.tv to the blacklist so that it at least does not impede your ability to use the website normally.

**Comment by Kushurando** [comment]
so, I looked at the experimental feature you linked here, and it doesn't show a second black list for me, did you merge them into a single blacklist?

**Comment by ChadBailey** [comment]
There is [definitely](https://github.com/igrigorik/videospeed/pull/547/files#diff-523e1fab8c682414ee02f4a211ccb4c7R123) a second blacklist in the options menu if you have the right version of the code.

**Comment by Kushurando** [comment]
oops, you're right, I downloaded the code but forgot to actually switch the extensions in my browser

**Comment by ChadBailey** [comment]
Unfortunately, incorporating this fix is blocked by #595 :(. I'm not sure how to proceed right now, but hope to find time over the next few days to take a closer look.

**Comment by N17m** [comment]
I would like to use the experimental feature discussed above which fixes the Twitch issue but I am new to github and I don't know what i'm doing. If this is something that can easily be explained then I would very much appreciate your help. If not then I understand and thank you for your help making/improving this extension regardless.

**Comment by jedawson** [comment]
You have to checkout his branch from his fork of the master repository, and
then load the extension into chrome after creating a new person in chrome
and turning on developer mode in that new person chrome window.

**Comment by ChadBailey** [comment]
After looking further into this, the hypothetical scenario I came up with was not validated. Any time you pause the video on twitch, resuming it returns you to the current time and there is no ability to seek through the backlog. I see the value in YouTube videos (which work), but I do not see the value on twitch.tv.

I suspect the users impacted are only using this as a mechanism for bypassing advertisements on Twitch. I'm of the opinion that using Video Speed Controller as a means to bypass advertisements is an antipattern, and I will not support bugfixes for the sole purpose of this antipattern.

Once the blocking issues are out of the way, Twitch.tv should be added to the default blacklist in order to improve user experience.

**Comment by Kushurando** [comment]
personally I just have an issue with the distorted audio when twitch tries to catchup to the vod, when you have built up buffer, by playing at 1.03 speed, and if you pause + play it sometimes doesn't reset to 1.00 requiring a full reload, which is annoying

**Comment by ChadBailey** [comment]
@Kushurando thanks for the clarification again. Adding twitch.tv to the blacklist will resolve the issue.

**Comment by N17m** [comment]
@jedawson Thank you for the reply but after spending about an hour of watching videos and googling some of the terms you used i seem no closer to understanding how to actually do any of the things you mentioned. I may have understated how clueless i am here :P. 

And @ChadBailey I can definitely understand your concerns regarding the extension being used improperly to bypass advertisements. I would say 90% of the time i am using the extension on twitch is to just go back in time and rewatch something, which still works fine and is not part of the issue discussed above. 

However, after going back i used to speed up the video to return to the live part of the stream, and i would also sometimes slow down the video when rewatching to analyze things more carefully. This is what i use the extension for on twitch most of the time. And technically all of this can still be done using the non live video recordings of the streams that are being recorded as the stream broadcasts(these are called VODs on twitch), it's just much much more convenient to do it on the fly in the live stream window with the extension. 

I see now though that it seems you were trying to make twitch work more like youtube livestreams where you can pause and play and scrub through a timeline on the livestream and perhaps make a coffee as well(i believe in you lol). Although that would be awesome im not really sure if that's even possible, having used the extension quiet a bit in the way i described above i can say there are some serious limitations to going back in time during a livestream, you can only go back so far, about 2-3min max, and even then if you go back too far the stream will begin jumping ahead multiple seconds at a time until it has caught up to the live broadcast. But i don't know anything about coding so *shrug*.

And returning to your point about the advertisement bypass, although i can definitely understand your concerns regarding this, especially considering what you described as the "inevitable outcome of this", i just really don't see anyone using this extension on twitch in place of an ad blocker, which works totally fine on twitch and is superior in every way.

Regardless, the twitch issue is a very small inconvenience for a benefit i didn't even know i wanted until way back when i downloaded this extension to watch youtube videos with more control and faster than the 2x speed limit, which is where i use the extension most, and only then did i notice that it worked on twitch as well. but you never know how much you miss something till it's gone i guess :P.

Forgive me for my wall of text and run on sentences.

**Comment by ChadBailey** [comment]
@N17m no apologies needed! I think I am the one who owes an apology here, I've clearly misunderstood the usage and concern. Open discussions like these are how we come together and find solutions!

Can I have you explain to me how I can reproduce the usage you describe? When testing I found that I was unable to do what I suspected in my mind to be able to do - but this was only based on guessing. I really would like some step by step instructions so I can be sure I'm understanding correctly.

**Comment by ChadBailey** [comment]
On second thought, since my experimental fix appears to be working I feel this may be unnecessary. I've got an open PR that resolves the blocking issue. Once that goes through I will get that experimental issue re-implemented as a resolution.

Thanks so much for your patience everyone!!!

**Comment by N17m** [comment]
I am actually pretty sure your most recent comment means you actually don't need me to explain anything anymore, in which case, feel free to ignore this.

But assuming that in your most recent comment where you say "On second thought... this may be unnecessary" you're not referring to your request for me to "explain to me how I can reproduce the usage you describe", then It's a bit unclear which part of what i said you want me to help you reproduce so i'll just explain the things i described.

In case the part you were confused about was when i said:
 "after going back i used to speed up the video to return to the live part of the stream, and i would also sometimes slow down the video when rewatching to analyze things more carefully"

I'll try to explain that, When watching a stream and something worth rewatching happens, i will:
1. Press the extensions "rewind" button to see it again.
Note: this part still works fine.
2. Press the "decrease speed" keybind to watch it in slow motion if i want.
3. Press the "increase speed" keybind until i reach about 1.5 speed when im done watching it.
4. Let the stream play until i am caught up to the live broadcast so that i don't miss anything. 

So steps 2-4 are what no longer work and from what i can tell are what you have fixed in your experimental feature.(which i still don't know how to use lol)

when i said:
   "you can only go back so far, about 2-3min max, and even then if you go back too far the stream 
   will begin jumping ahead multiple seconds at a time until it has caught up to the live broadcast."

basically i was just wrong, let me try to revise this: 
You can only go back about 30 seconds before the stream forces you back to live video. 
And if you rewind and rewatch the same thing over and over for long enough then when you try to let the video play normally at some point it will skip ahead multiple seconds of video. 

To help see the buffer size before you try to reproduce this, click the settings cog icon in the bottom right of the twitch streams video, then click "advanced", then check the "Video stats" setting. Now you can see the "Buffer size" in the window that should have appeared in the top left.

for the "you can only go back 30 seconds" part
you can reproduce this by: 
1.   Open a twitch stream.
2.   let the stream play for at least 1 minutes so we have some video to work with.
3.   now just use the "Rewind" feature from the extension to go back as far as you can by pressing the the rewind keybind over and over.

you should see that you can only go back about 30 seconds before it just forces you back to live and stops you from going back altogether until you load more video.

for the rewinding and rewatching until the video skips ahead part
you can reproduce this by:  
1. Rewind video until the buffer size says 30 seconds.
(if you do this too quickly you will be sent back to live as described earlier so make sure you add about a half a second between each press.)
Note: At this point you could happily watch the stream with basically a 30 second delay.
2. if you let the stream play for the same amount of time that your extensions Rewind time is set to then press the rewind button and repeat.(basically rewatching the same few seconds over and over) When you stop doing this and let the video play normally, after letting it play normally for about 30 seconds you will see it skip ahead several seconds. 

And to be clear, im not asking you to fix or improve any of the things i'm describing here, in case it seemed like i was in either this comment or any of my earlier comments.

**Comment by ChadBailey** [comment]
Your explanations are very valuable, I can see how this truly enhances your experience when using Twitch.

Based on what @Kushurando expressed, I do think the experimental fix will resolve the issues and let you resume using it on Twitch.

**Comment by ChadBailey** [comment]
Thanks so much for everyone's great feedback and conversations. I have finally implemented this fix (#604) using the new code base that should work in shadowroot contexts.

Hopefully it gets merged soon and you can all return to using twitch the way you used to.
